File: Sources/CodexQuotaCore/QuotaEventDecoder.swift
import Foundation

public struct DecodedRateLimitEvent: Equatable, Sendable {
    public struct Window: Equatable, Sendable {
        public let usedPercent: Double
        public let windowMinutes: Int
        public let resetsAt: Int64

        init(usedPercent: Double, windowMinutes: Int, resetsAt: Int64) {
            self.usedPercent = usedPercent
            self.windowMinutes = windowMinutes
            self.resetsAt = resetsAt
        }
    }

    public let observedAt: Date
    public let limitID: String
    public let primary: Window?
    public let secondary: Window?
}

public enum QuotaEventDecoder {
    public static func decode(line: String) -> DecodedRateLimitEvent? {
        guard
            let data = line.data(using: .utf8),
            let envelope = try? JSONDecoder().decode(Envelope.self, from: data),
            envelope.type == "event_msg",
            let rateLimits = envelope.payload.rateLimits,
            let observedAt = parseTimestamp(envelope.timestamp)
        else {
            return nil
        }

        return .init(
            observedAt: observedAt,
            limitID: rateLimits.limitID,
            primary: rateLimits.primary.map(makeWindow),
            secondary: rateLimits.secondary.map(makeWindow)
        )
    }

    private static func makeWindow(_ window: Envelope.DecodedWindow) -> DecodedRateLimitEvent.Window {
        .init(
            usedPercent: window.usedPercent,
            windowMinutes: window.windowMinutes,
            resetsAt: window.resetsAt
        )
    }

    private static func parseTimestamp(_ value: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.date(from: value)
    }

    private struct Envelope: Decodable {
        let timestamp: String
        let type: String
        let payload: Payload

        struct Payload: Decodable {
            let rateLimits: RateLimits?

            enum CodingKeys: String, CodingKey {
                case rateLimits = "rate_limits"
            }
        }

        struct RateLimits: Decodable {
            let limitID: String
            let primary: DecodedWindow?
            let secondary: DecodedWindow?

            enum CodingKeys: String, CodingKey {
                case limitID = "limit_id"
                case primary
                case secondary
            }
        }

        struct DecodedWindow: Decodable {
            let usedPercent: Double
            let windowMinutes: Int
            let resetsAt: Int64

            enum CodingKeys: String, CodingKey {
                case usedPercent = "used_percent"
                case windowMinutes = "window_minutes"
                case resetsAt = "resets_at"
            }
        }
    }
}

File: windows/CodexQuota.Windows/QuotaEventDecoder.cs
using System.Globalization;
using System.Text.Json;

namespace CodexQuota.Windows;

public static class QuotaEventDecoder
{
    public static DecodedRateLimitEvent? Decode(string line)
    {
        try
        {
            using var document = JsonDocument.Parse(line, new JsonDocumentOptions { MaxDepth = 32 });
            var root = document.RootElement;
            if (!TryGetString(root, "type", out var type) || type != "event_msg" ||
                !TryGetString(root, "timestamp", out var timestamp) ||
                !DateTimeOffset.TryParse(
                    timestamp,
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.AllowWhiteSpaces,
                    out var observedAt) ||
                !root.TryGetProperty("payload", out var payload) ||
                payload.ValueKind != JsonValueKind.Object ||
                !payload.TryGetProperty("rate_limits", out var rateLimits) ||
                rateLimits.ValueKind != JsonValueKind.Object ||
                !TryGetString(rateLimits, "limit_id", out var limitId) ||
                !TryDecodeWindow(rateLimits, "primary", out var primary) ||
                !TryDecodeWindow(rateLimits, "secondary", out var secondary))
            {
                return null;
            }

            return new DecodedRateLimitEvent(
                observedAt,
                limitId,
                primary,
                secondary);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static bool TryDecodeWindow(JsonElement rateLimits, string name, out QuotaWindow? window)
    {
        window = null;
        if (!rateLimits.TryGetProperty(name, out var value) || value.ValueKind == JsonValueKind.Null)
        {
            return true;
        }
        if (value.ValueKind != JsonValueKind.Object ||
            !value.TryGetProperty("used_percent", out var usedPercent) || !usedPercent.TryGetDouble(out var used) ||
            !value.TryGetProperty("window_minutes", out var windowMinutes) || !windowMinutes.TryGetInt32(out var minutes) ||
            !value.TryGetProperty("resets_at", out var resetsAt) || !resetsAt.TryGetInt64(out var reset))
        {
            return false;
        }

        window = new QuotaWindow(used, minutes, reset);
        return true;
    }

    private static bool TryGetString(JsonElement value, string name, out string result)
    {
        result = string.Empty;
        if (!value.TryGetProperty(name, out var property) || property.ValueKind != JsonValueKind.String)
        {
            return false;
        }

        result = property.GetString() ?? string.Empty;
        return true;
    }
}
